### PR TITLE
EOS-20132: Health data for disks and psus

### PR DESCRIPTION
**Pull request by sandeepsgt**
## Coding:

* [x] Did you follow coding standard and verified it with flake8? 

* [x] Did your address all Codacy issues?

## Testing:

* [ ] Unittest updated and all unittests are passing?

* [ ] Sanity/Self tests are updated (if applicable)?

* [ ] Manual testing done covering happy path and non-happy path?

## Integration:

* [ ] If there is any interface change, did you communicate it to other components?
* [ ] If there is any interface change, other componenent gate gatekeepers are ready to accept the change?
* [ ] Did you complete deployment test if applicable?

## PR checklist:
* [x] Did you add Jira number to the PR?

* [x] DCO and cla-signed?

### CODE REVIEW

File: low-level/files/opt/seagate/sspl/setup/resource_map/server.py [hunk 1]
```diff
@@ -32,18 +42,365 @@ class ServerMap(ResourceMap):
...
+            return value.decode('utf-8')
+        else:
+            # dbus.Array([dbus.Byte]) or any similar sequence type:
+            return bytearray(value).rstrip(bytearray((0,))).decode('utf-8')
+
 
+    def get_psu_info(self):
+        """Update and return PSU information in specific format."""
+        psus = []
+        response, _, _ = SimpleProcess("dmidecode").run()
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/resource_map/server.py:392]: Can limit to psu type

>]# dmidecode -t 39
# dmidecode 3.2
Getting SMBIOS data from sysfs.
SMBIOS 3.2.1 present.
# SMBIOS implementations newer than version 3.2.0 are not
# fully supported by this version of dmidecode.

Handle 0x000C, DMI type 39, 22 bytes
System Power Supply
        Power Unit Group: 1
        Location: PSU1
        Name: PWS-751P-1R
        Manufacturer: SUPERMICRO
        Serial Number: P751PCJ31CU1144
        Asset Tag: N/A
        Model Part Number: PWS-751P-1R
        Revision: 1.1
        Max Power Capacity: 750 W
        Status: Present, OK
        Type: Switching
        Input Voltage Range Switching: Auto-switch
        Plugged: Yes
        Hot Replaceable: Yes

Handle 0x000D, DMI type 39, 22 bytes
System Power Supply
        Power Unit Group: 2
        Location: PSU2
        Name: PWS-751P-1R
        Manufacturer: SUPERMICRO
        Serial Number: P751PCJ31CU1107
        Asset Tag: N/A
        Model Part Number: PWS-751P-1R
        Revision: 1.1
        Max Power Capacity: 750 W
        Status: Present, OK
        Type: Switching
        Input Voltage Range Switching: Auto-switch
        Plugged: Yes
        Hot Replaceable: Yes

File: low-level/files/opt/seagate/sspl/setup/resource_map/server.py [hunk 2]
```diff
@@ -32,18 +42,365 @@ class ServerMap(ResourceMap):
...
+            self.set_health_data(cpu_dict, status=health_status,
+                                 specification=specification)
+            cpu_data.append(cpu_dict)
+        return cpu_data
+
+    def get_disks_info(self):
+        """Update and return server drive information in specific format."""
+        return self._get_drives()
+
+    def _get_drives(self):
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/resource_map/server.py:228]: seems code is duplicated from disk sensor. Better would be to add these common functions in common utilities and use from there, so later sensor also we can  update to use from there

**Review by thavanathan — COMMENTED** [review]
@sandeepsgt please fill PR template.

File: low-level/framework/utils/server_drives.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+# about this software or licensing, please email opensource@seagate.com or
+# cortx-questions@seagate.com.
+
+
+import json
+from dbus import Array, Interface, SystemBus
+from cortx.utils.process import SimpleProcess
+import re
+
+class ServerDrives:
```
Comment (cdeshmukh) [inline low-level/framework/utils/server_drives.py:22]: consider renaming class name, as we discussed in design discussion

File: low-level/framework/utils/server_drives.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
+# Copyright (c) 2001-2020 Seagate Technology LLC and/or its Affiliates
```
Comment (cdeshmukh) [inline low-level/framework/utils/server_drives.py:1]: Can we put this under 
low-level/framework/platform/server/drives

Lets start building the hierarchy classes

File: low-level/files/opt/seagate/sspl/setup/resource_map/server.py [hunk 5]
```diff
@@ -269,9 +275,51 @@ def prepare_mem_json(self, status, description):
...
+        }]
+        return disks_health_info
+
+
+    def get_psu_info(self):
+        """Update and return PSU information in specific format."""
+        psus = []
+        response, _, _ = SimpleProcess("dmidecode -t 39").run()
+        response = response.decode()
+        matches = re.finditer("System Power Supply\n(\s*.*\n)\s*Location: (?P<location>.*)\n(\s*.*\n){7}\s*Status: (?P<status>.*)\n(\s*.*\n){2}\s*Plugged: (?P<plugged>.*)", response)
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/resource_map/server.py:312]: limit this to 80 chars length

File: low-level/framework/utils/server_drives.py (hunk 3 above)
Comment (sandeepsgt) [reply low-level/framework/utils/server_drives.py]: Done

File: low-level/files/opt/seagate/sspl/setup/resource_map/server.py [hunk 6]
```diff
@@ -269,9 +275,51 @@ def prepare_mem_json(self, status, description):
...
+            self.set_health_data(data, health, specifics=psu_data)
+            psus.append(data)
+        return psus
 
-# if __name__ == "__main__":
-#     server = ServerMap()
-#     health_data = server.get_health_info(
-#         rpath="nodes[0]>compute[0]>hw>memory")
-#     print(health_data)
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/server.py]: Done

File: low-level/files/opt/seagate/sspl/setup/resource_map/server.py (hunk 5 above)
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/server.py]: Done

File: low-level/framework/utils/server_drives.py (hunk 4 above)
Comment (sandeepsgt) [reply low-level/framework/utils/server_drives.py]: Done

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py [hunk 7]
```diff
@@ -461,3 +468,56 @@ def get_nw_status(self, nw_interface, interface):
...
+                "diskUsedPercentage": psutil.disk_usage("/")[3],
+            }
+        disks = []
+        for disk in Disks().get_disks():
+            uid = disk.path if disk.path else disk.id
+            disk_health = self.get_health_template(uid, True)
+            health_data = disk.get_health()
+            health = "OK" if health_data['SMART_health'] else "Fault"
+            self.set_health_data(disk_health, health, specifics=[{"SMART": health_data}])
+            disk_health["last_updated"] = int(time.time())
```
Comment (shriya-deshmukh) [inline low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:488]: set_health_data function  takes care of  adding 'last_updated' filed in the schema, can remove this line.

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py [hunk 8]
```diff
@@ -461,3 +468,56 @@ def get_nw_status(self, nw_interface, interface):
...
+        response, _, _ = SimpleProcess("dmidecode -t 39").run()
+        matches = re.finditer(r"System Power Supply\n(\s*.*\n)"
+                              r"\s*Location: (?P<location>.*)\n"
+                              r"(\s*.*\n){7}\s*Status: (?P<status>.*)\n"
+                              r"(\s*.*\n){2}\s*Plugged: (?P<plugged>.*)",
+                              response.decode())
+        for match in matches:
+            psu_data = match.groupdict()
+            data = self.get_health_template(f'Power Supply {uids[psu_data["location"]]}', True)
+            data["last_updated"] = int(time.time())
```
Comment (shriya-deshmukh) [inline low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:519]: set_health_data function takes care of adding 'last_updated' filed in the schema, can remove this line.

File: low-level/framework/platforms/server/disk.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+import json
+import re
+
+from cortx.utils.process import SimpleProcess
+from dbus import Array, Interface, SystemBus
+
+
+class Disk:
+
+    def __init__(self, id, path, device) -> None:
```
Comment (cdeshmukh) [inline low-level/framework/platforms/server/disk.py]: >Redefining built-in 'id'

check this codacy warning

File: low-level/framework/platforms/server/disk.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+            # "SG_IO: bad/missing sense data") for local drives,
+            # this check would fail.
+            if DISK_ERR_MISSING_SENSE_DATA not in err and \
+                DISK_ERR_GET_ID_FAILURE not in err:
+                return True
+            else:
+                return False
+
+
+class Disks:
```
Comment (cdeshmukh) [inline low-level/framework/platforms/server/disk.py:92]: why 2nd class? function names n description can differentiate their intent, atlast resource type is same

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py [hunk 11]
```diff
@@ -461,3 +468,56 @@ def get_nw_status(self, nw_interface, interface):
...
+        units_factor_GB = 1000000000
+        overall_usage = {
+                "totalSpace": f'{int(psutil.disk_usage("/")[0])//int(units_factor_GB)} GB',
+                "usedSpace": f'{int(psutil.disk_usage("/")[1])//int(units_factor_GB)} GB',
+                "freeSpace": f'{int(psutil.disk_usage("/")[2])//int(units_factor_GB)} GB',
+                "diskUsedPercentage": psutil.disk_usage("/")[3],
+            }
+        disks = []
+        for disk in Disks().get_disks():
+            uid = disk.path if disk.path else disk.id
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:483]: >https://jts.seagate.com/secure/attachment/467631/disks.json
        "uid": "ST1000NM0055-1V410C-ZC236QHZ",

we need disk_\<slot> or disk dev-id with path kind of uid, serial number can be in specifics

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py [hunk 12]
```diff
@@ -461,3 +468,56 @@ def get_nw_status(self, nw_interface, interface):
...
+        return disks_health_info
+
+    def get_psu_info(self):
+        """Update and return PSU information in specific format."""
+        psus = []
+        psu_mapping = {
+            "PS1 Status": "PSU1",
+            "PS2 Status": "PSU2"
+        }
+        response, _, _ = SimpleProcess(shlex.split('ipmitool sdr type "Power Supply" -c')).run()
```
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:505]: shell cmd lexer  is added to Simple Process now @sandeepsgt , so no need to lex the cmd now, we can directly pass the string `'ipmitool sdr type "Power Supply" -c'`

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py [hunk 13]
```diff
@@ -16,22 +16,27 @@
...
+import re
+import shlex
 import time
-from socket import AF_INET
 from pathlib import Path
+from socket import AF_INET
 
+import psutil
+from cortx.utils.process import SimpleProcess
+from framework.base.sspl_constants import (CPU_PATH, DEFAULT_RECOMMENDATION,
```
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:27]: this way it asks for a lot of indentation going forward if we have to add more imports,
as earlier suggested by @mariyappanp, following is the better option.
```
from framework.base.sspl_constants import (
    CPU_PATH, DEFAULT_RECOMMENDATION, HEALTH_SVC_NAME, SAS_RESOURCE_ID)
```

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 12 above)
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:505]: why are we using direct ipmitool cmd, @sandeepsgt  @cdeshmukh, we should be using IPMI_Client right??

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 7 above)
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py]: Done

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 8 above)
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py]: Done

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 12 above)
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py]: Used IPMI Client class

**Review by shriya-deshmukh — APPROVED** [review]

**Review by sumedhak27 — APPROVED** [review]

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py [hunk 14]
```diff
@@ -461,3 +468,56 @@ def get_nw_status(self, nw_interface, interface):
...
+            "PS1 Status": "PSU1",
+            "PS2 Status": "PSU2"
+        }
+        response, _, _ = SimpleProcess(shlex.split('ipmitool sdr type "Power Supply" -c')).run()
+        uids = {}
+        for psu in response.decode().strip("\n").split("\n"):
+            psu_name, sensor, *_ = psu.split(",")
+            uids[psu_mapping[psu_name]] = f"#0x{sensor.strip('h').lower()}"
+        response, _, _ = SimpleProcess("dmidecode -t 39").run()
+        matches = re.finditer(r"System Power Supply\n(\s*.*\n)"
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:509]: We have so many fields from dmidecode, why we are not using these 
>System Power Supply
        Location: PSU1
        Name: PWS-751P-1R
        Manufacturer: SUPERMICRO
        Serial Number: P751PCJ32CU1284
        Revision: 1.1
        Max Power Capacity: 750 W
        Status: Present, OK
        Type: Switching
        Input Voltage Range Switching: Auto-switch
        Plugged: Yes
        Hot Replaceable: Yes

**Review by cdeshmukh — COMMENTED** [review]
https://jts.seagate.com/secure/attachment/467631/disks.json

generated data doesnt seem to comply to schema, which will not reflect this data in final o/p then
@mariyappanp guess this has same issue as CPU

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py [hunk 15]
```diff
@@ -461,3 +468,53 @@ def get_nw_status(self, nw_interface, interface):
...
+                              r"\s*Location: (?P<location>.*)\n"
+                              r"(\s*.*\n){7}\s*Status: (?P<status>.*)\n"
+                              r"(\s*.*\n){2}\s*Plugged: (?P<plugged>.*)",
+                              response.decode())
+        for match in matches:
+            psu_data = match.groupdict()
+            data = self.get_health_template(f'Power Supply {uids[psu_data["location"]]}', True)
+            health = "OK" if (psu_data["status"] == "Present, OK") else "Fault"
+            self.set_health_data(data, health, specifics=psu_data)
+            psus.append(data)
```
Comment (gargpalak048) [inline low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:519]: add debug log to record the generated data as doing for other modules

**Review by gargpalak048 — CHANGES_REQUESTED** [review]

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py [hunk 16]
```diff
@@ -16,22 +16,27 @@
 # please email opensource@seagate.com or cortx-questions@seagate.com
 
 import errno
-import psutil
+import re
+import shlex
```
Comment (gargpalak048) [inline low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:20]: > import shlex

I think shlex is not getting used anywhere.
can remove this import

**Review by gargpalak048 — CHANGES_REQUESTED** [review]

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 14 above)
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py]: Used below fields now

```
       "Power Unit Group": "2",
        "Location": "PSU2",
        "Name": "PWS-751P-1R",
        "Serial Number": "P751PCJ35CU5374",
        "Max Power Capacity": "750 W",
        "Status": "Present, OK",
        "Plugged": "Yes",
        "Hot Replaceable": "Yes"
```

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 11 above)
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:502]: Used by-path
ex. pci-0000:00:11.5-ata-2.0

File: low-level/framework/platforms/server/disk.py (hunk 10 above)
Comment (sandeepsgt) [reply low-level/framework/platforms/server/disk.py:92]: Disk object represent one disk( health , id etc)
Disks class creates Disk object from dbus

File: low-level/framework/platforms/server/disk.py (hunk 9 above)
Comment (sandeepsgt) [reply low-level/framework/platforms/server/disk.py]: Done

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 15 above)
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py]: Done

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 16 above)
Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py]: Done

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 11 above)
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:502]: >ex. pci-0000:00:11.5-ata-2.0

thats how it looks for a disk ?

Comment (sandeepsgt) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:502]: We are using same in sensor.

File: low-level/framework/platforms/server/disk.py (hunk 10 above)
Comment (cdeshmukh) [reply low-level/framework/platforms/server/disk.py:92]: class names not giving that differentiation !
why cant we merge these two classes and have only 'Disk' and let dbus related functions stay in that too

or name Disks -> something distinctive and a noun still

File: low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py (hunk 11 above)
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/resource_map/lr2/server.py:502]: hmm so in sensor also we would need the change. keep TODO

**Review by cdeshmukh — APPROVED** [review]

**Review by jjxsg — APPROVED** [review]

**Review by cdeshmukh — APPROVED** [review]

**Review by jjxsg — APPROVED** [review]

**Pull request merged**